File: Leetcode/Problems C++/695. Max Area of Island.cpp
class Solution {
public:
    int findArea(vector<vector<int>>& grid, int nx,int ny, int x,int y)
    {
        if(x<0 || y<0 || x>=nx || y>=ny)
            return 0;
        if(grid[x][y]==0)
            return 0;
        grid[x][y]=0;
        return 1+findArea(grid,nx,ny,x-1,y)+findArea(grid,nx,ny,x+1,y)+findArea(grid,nx,ny,x,y+1)+findArea(grid,nx,ny,x,y-1);
    }
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int max_area= 0;
        int nx= grid.size();
        int ny= grid[0].size();
        for(int i=0;i<nx;i++)   //NOT from 0 as [[1]]
        {
            for(int j=0;j<ny;j++)
            {
                if(grid[i][j]==1)
                {
                    max_area = max(max_area,findArea(grid,nx,ny,i,j));
                }
            }
        }
        return max_area;
    }
};

File: Leetcode/Problems Java/695. Max Area of Island.java
class Solution {
    //O(n)
    public int maxAreaOfIsland(int[][] grid) {
        int n=grid.length;
        int m=grid[0].length;
        int ans = 0;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==1){
                    ans = Math.max(ans, dfs(grid,i,j,n,m));
                }
            }
        }
        return ans;
    }
    private int dfs(int [][]grid, int i,int j,int n, int m){
        if(i<0 || i>=n || j<0 || j>=m || grid[i][j]!=1)
            return 0;
        grid[i][j]=-1;
        return 1 + dfs(grid,i-1,j,n,m)+dfs(grid,i+1,j,n,m)+dfs(grid,i,j-1,n,m)+dfs(grid,i,j+1,n,m);
    }
}
